pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/table/internal/SymbolTableResolver.java

```diff
@@ -213,11 +213,19 @@ public Void visit(ASTCompilationUnit node, @NonNull ReferenceCtx ctx) {
             pushed += pushOnStack(f.samePackageSymTable(top()));
             pushed += pushOnStack(f.singleImportsSymbolTable(top(), isImportOnDemand.get(false)));
 
+            // TODO Java 23 implicitly imports "import static java.io.IO.*" for implicitly declared classes
+            boolean implicitlyDeclaredClass = node.isImplicitlyDeclaredClass();
+
             NodeStream<ASTTypeDeclaration> typeDecls = node.getTypeDeclarations();
 
             // types declared inside the compilation unit
             pushed += pushOnStack(f.typesInFile(top(), typeDecls));
 
+            if (implicitlyDeclaredClass) {
```

Comment:
I think we need to add a node that implements ASTAnyTypeDeclaration for the implicit class declaration. Otherwise the method "getEnclosingType" that is available on all nodes may return null in places where we expect there to be an enclosing class (eg methods or field declarations), which is a source of bugs.

These if statements won't be necessary anymore.

Reply:
I've added now a new node "ImplicitClassDeclaration". Its structure is similar to an anonymous class declaration.